### DISCUSSION TITLE
Desktop Java usage after removal in v1.11.0

### DISCUSSION THREAD

**Discussion by darnells** [google/filament#4661] [Q&A] [answered]
Hello, I understand that Desktop Java support was removed in the v1.11.0 release. The description as to why is self explanatory in the comments for pull #4263

> These builds are never tested nor used on our end. We cannot
> guarantee their proper support. It should also be possible
> for an app to handle this itself.

Before this, we were working with v1.9.25 to construct a 3D molecular viewer in a RCP/SWT Java application on Windows and macOS. Even before the removal, our initial thoughts were to create the canvas in Java and use a native C++ library to perform the rendering actions triggered through JNI or JNA calls.

Moving forward, what would be the advice "for an app to handle [Desktop Java bindings] itself"? Would our proposed scheme significantly affect that advice? Would the reintroduction of the Desktop Java bindings be considered if there was a commitment to aid with their testing and support?

Thank you for your advice as well as providing a great tool!

**Comment by romainguy** [comment, marked as answer]
Your proposal wouldn't change our advice. A good starting point would be our former [Java bindings](https://github.com/google/filament/tree/77891acb9211855420e62e2f81e1fd9bc185e9db/java/filament/src/java/com/google/android/filament) as well as the [required JNI setup](https://github.com/google/filament/tree/77891acb9211855420e62e2f81e1fd9bc185e9db/android/filament-android/src/main/cpp/nativewindow). It's not a lot of code and should be fairly easy to integrate in your application.

The only "trick" is that Java2D expects ARGB pixel data and Filament outputs RGBA so there's code in the Java bindings to swizzle the framebuffer. It would be better to do it directly on the GPU though.
